go/vt/mysqlctl/compression.go

```diff
@@ -0,0 +1,313 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysqlctl
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os/exec"
+	"sync"
+
+	"github.com/google/shlex"
+	"github.com/klauspost/compress/zstd"
+	"github.com/klauspost/pgzip"
+	"github.com/pierrec/lz4"
+	"github.com/planetscale/pargzip"
+
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+var (
+	compressionLevel = flag.Int("compression_level", 1, "what level to pass to the compressor")
+	// switch which compressor/decompressor to use
+	BuiltinCompressor   = flag.String("builtin_compressor", "pgzip", "builtin compressor engine to use")
+	BuiltinDecompressor = flag.String("builtin_decompressor", "auto", "builtin decompressor engine to use")
+	// use and external command to decompress the backups
+	ExternalCompressorCmd   = flag.String("external_compressor", "", "command with arguments to use when decompressing a backup")
+	ExternalCompressorExt   = flag.String("external_compressor_extension", "", "extension to use when using an external decompressor")
+	ExternalDecompressorCmd = flag.String("external_decompressor", "", "command with arguments to use when compressing a backup")
+
+	errUnsupportedCompressionEngine    = errors.New("unsupported engine")
+	errUnsupportedCompressionExtension = errors.New("unsupported extension")
+
+	// this is used by getEngineFromExtension() to figure out which engine to use in case the user didn't specify
+	engineExtensions = map[string][]string{
+		".gz":  {"pgzip", "pargzip"},
+		".lz4": {"lz4"},
+		".zst": {"zstd"},
+	}
+)
+
+func getEngineFromExtension(extension string) (string, error) {
+	for ext, eng := range engineExtensions {
+		if ext == extension {
+			return eng[0], nil // we select the first supported engine in auto mode
+		}
+	}
+
```

Comment:
We have lots of unnecessary newlines in these new files. Can you remove them?

Reply:
I don't find many tbh. I have removed some but let me know if you see any specific instances.